Feature: Add & syntax for mutable function parameters

## Summary

Add explicit `&` syntax to declare mutable function parameters. By default, function parameters should be read-only. The `&` prefix opts into allowing modification of the parameter's internal data (struct fields, array elements, map entries).

## Motivation

Currently, there is inconsistent behavior:
- Struct fields can be modified through function parameters (unintentional)
- Array elements cannot be modified through function parameters
- Map entries cannot be modified through function parameters

Additionally, `const` variables can be modified through function calls, violating the expectation that `const` means immutable.

## Proposed syntax

```
// & means "I will modify this parameter"
do birthday(&p Person) {
    p.age = p.age + 1  // OK
}

// No symbol means "read-only"  
do get_name(p Person) -> string {
    // p.age = 100  // ERROR: cannot modify read-only parameter
    return p.name
}
```

## Rules

| Param declaration | Can modify inside function? |
|-------------------|---------------------------|
| `p Person` | No |
| `&p Person` | Yes |

| Caller variable | To `p Person` | To `&p Person` |
|-----------------|---------------|----------------|
| `temp` | OK (read-only) | OK (writable) |
| `const` | OK (read-only) | **ERROR** |

## Error messages

New error codes will need to be added:

**E5015 - Attempting to modify read-only parameter:**
```
error[E5015]: cannot modify read-only parameter 'p'
  --> main.ez:3:5
   |
 2 | do get_name(p Person) -> string {
   |             - parameter is not declared as mutable
 3 |     p.age = 100
   |     ^^^^^ cannot modify
   |
help: add '&' to make parameter mutable: &p Person
```

**E5016 - Passing const to mutable parameter:**
```
error[E5016]: cannot pass immutable variable 'alice' to mutable parameter '&p'
  --> main.ez:10:10
   |
 5 | do birthday(&p Person) {
   |             -- parameter requires mutable argument
   |
10 |     birthday(alice)
   |              ^^^^^ 'alice' is declared as const
   |
help: either declare 'alice' as temp, or remove & from parameter
```

## Implementation

| Component | Change |
|-----------|--------|
| Lexer | Recognize `&` in parameter context |
| Parser | Parse `&` prefix on parameter names |
| AST | Add `Mutable bool` field to parameter nodes |
| Type checker | Error if passing `const` to `&` param |
| Type checker | Error if modifying non-`&` param |
| Evaluator | Set param mutability from declaration |
| Evaluator | Check mutability on struct field assignment |
| Errors | Add new error codes E5015, E5016 (or appropriate codes) |

## Breaking change

This changes default behavior. Existing code that modifies struct fields in functions will need to add `&` to those parameters:

```
// Before (worked due to bug)
do birthday(p Person) {
    p.age = p.age + 1
}

// After (explicit)
do birthday(&p Person) {
    p.age = p.age + 1
}
```

## Design rationale

1. **Explicit over implicit** — Function signature clearly shows intent to modify
2. **No new keywords** — Single `&` symbol, minimal syntax addition
3. **Consistent** — Same rules for structs, arrays, and maps
4. **Compile-time safety** — Errors caught before runtime
5. **Protects const** — `const` variables cannot be accidentally modified through function calls

--- a/.gitignore
+++ b/.gitignore
@@ -4,6 +4,7 @@ dist/
 context
 *.txt
 tmp/
+.DS_Store
 
 # Ignore all .ez files by default (for local scratch/testing)
 *.ez

--- a/pkg/ast/ast.go
+++ b/pkg/ast/ast.go
@@ -417,6 +417,7 @@ func (f *FunctionDeclaration) TokenLiteral() string { return f.Token.Literal }
 type Parameter struct {
 	Name     *Label
 	TypeName string
+	Mutable  bool // true if declared with & prefix
 }
 
 // ImportItem represents a single module import with optional alias

--- a/pkg/errors/codes.go
+++ b/pkg/errors/codes.go
@@ -96,6 +96,7 @@ var (
 	E3020 = ErrorCode{"E3020", "negative-to-unsigned", "cannot assign negative value to unsigned type"}
 	E3021 = ErrorCode{"E3021", "byte-value-out-of-range", "byte value must be between 0 and 255"}
 	E3022 = ErrorCode{"E3022", "byte-array-element-out-of-range", "byte array element must be between 0 and 255"}
+	E3023 = ErrorCode{"E3023", "const-to-mutable-param", "cannot pass immutable variable to mutable parameter"}
 )
 
 // =============================================================================
@@ -134,6 +135,7 @@ var (
 	E5013 = ErrorCode{"E5013", "range-start-not-integer", "range start must be integer"}
 	E5014 = ErrorCode{"E5014", "range-end-not-integer", "range end must be integer"}
 	E5015 = ErrorCode{"E5015", "postfix-requires-identifier", "postfix operator needs variable"}
+	E5016 = ErrorCode{"E5016", "immutable-parameter", "cannot modify read-only parameter"}
 )
 
 // =============================================================================

--- a/pkg/interpreter/evaluator.go
+++ b/pkg/interpreter/evaluator.go
@@ -780,6 +780,23 @@ func evalAssignment(node *ast.AssignmentStatement, env *Environment) Object {
 
 	switch target := node.Name.(type) {
 	case *ast.Label:
+		// Check if the variable is a Reference (for & params)
+		if existingVal, ok := env.Get(target.Value); ok {
+			if ref, isRef := existingVal.(*Reference); isRef {
+				// Handle compound assignment for references
+				if node.Operator != "=" {
+					oldVal, _ := ref.Deref()
+					val = evalCompoundAssignment(node.Operator, oldVal, val, node.Token.Line, node.Token.Column)
+					if isError(val) {
+						return val
+					}
+				}
+				// Update through the reference
+				ref.SetValue(val)
+				return NIL
+			}
+		}
+
 		// Handle compound assignment
 		if node.Operator != "=" {
 			oldVal, ok := env.Get(target.Value)
@@ -1306,6 +1323,12 @@ func evalFunctionDeclaration(node *ast.FunctionDeclaration, env *Environment) Ob
 
 func evalIdentifier(node *ast.Label, env *Environment) Object {
 	if val, ok := env.Get(node.Value); ok {
+		// If the value is a Reference (for & params), dereference it
+		if ref, isRef := val.(*Reference); isRef {
+			if derefVal, ok := ref.Deref(); ok {
+				return derefVal
+			}
+		}
 		// Mutability is now set at declaration time, not at lookup time
 		// This preserves the original object's mutability when passed to functions
 		return val
@@ -1788,12 +1811,47 @@ func evalCallExpression(node *ast.CallExpression, env *Environment) Object {
 		return function
 	}
 
+	// For user-defined functions with mutable (&) params, handle references specially
+	if fn, ok := function.(*Function); ok {
+		args := evalArgsWithReferences(node.Arguments, fn.Parameters, env)
+		if len(args) == 1 && isError(args[0]) {
+			return args[0]
+		}
+		return applyFunction(function, args, node.Token.Line, node.Token.Column)
+	}
+
+	// For builtins and other callables, evaluate arguments normally
 	args := evalExpressions(node.Arguments, env)
 	if len(args) == 1 && isError(args[0]) {
 		return args[0]
 	}
 
 	return applyFunction(function, args, node.Token.Line, node.Token.Column)
+}
+
+// evalArgsWithReferences evaluates arguments, creating References for mutable (&) params
+func evalArgsWithReferences(argExprs []ast.Expression, params []*ast.Parameter, env *Environment) []Object {
+	args := make([]Object, len(argExprs))
+
+	for i, argExpr := range argExprs {
+		// Check if this parameter is mutable and the argument is a variable
+		if i < len(params) && params[i].Mutable {
+			if label, ok := argExpr.(*ast.Label); ok {
+				// Create a reference to the original variable
+				args[i] = &Reference{Env: env, Name: label.Value}
+				continue
+			}
+		}
+
+		// Otherwise, evaluate normally
+		evaluated := Eval(argExpr, env)
+		if isError(evaluated) {
+			return []Object{evaluated}
+		}
+		args[i] = evaluated
+	}
+
+	return args
 }
 
 func evalMemberCall(member *ast.MemberExpression, args []ast.Expression, env *Environment) Object {
@@ -2081,7 +2139,8 @@ func extendFunctionEnv(fn *Function, args []Object) *Environment {
 
 	for i, param := range fn.Parameters {
 		if i < len(args) {
-			env.Set(param.Name.Value, args[i], false) // params are immutable (const)
+			// Use parameter's Mutable field: & params are mutable, non-& params are immutable
+			env.Set(param.Name.Value, args[i], param.Mutable)
 		}
 	}
 

--- a/pkg/interpreter/evaluator_test.go
+++ b/pkg/interpreter/evaluator_test.go
@@ -892,3 +892,188 @@ r
 	evaluated := testEval(input)
 	testIntegerObject(t, evaluated, 55)
 }
+
+// ============================================================================
+// Mutable Parameter Tests (& prefix)
+// ============================================================================
+
+func TestMutableStructParameter(t *testing.T) {
+	// Test that & param allows modification and changes persist
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do birthday(&p Person) {
+	p.age = p.age + 1
+}
+
+temp bob Person = Person{name: "Bob", age: 30}
+birthday(bob)
+bob.age
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 31)
+}
+
+func TestImmutableStructParameterNoChange(t *testing.T) {
+	// Test that non-& param doesn't modify the original (read-only copy behavior)
+	// Note: This test verifies the evaluator behavior - the typechecker would
+	// actually error on this code, but the evaluator should handle it gracefully
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do getName(p Person) -> string {
+	return p.name
+}
+
+temp bob Person = Person{name: "Bob", age: 30}
+temp n string = getName(bob)
+n
+`
+	evaluated := testEval(input)
+	testStringObject(t, evaluated, "Bob")
+}
+
+func TestMutableArrayParameter(t *testing.T) {
+	// Test that & param allows array modification
+	input := `
+do doubleFirst(&arr [int]) {
+	arr[0] = arr[0] * 2
+}
+
+temp nums [int] = {10, 20, 30}
+doubleFirst(nums)
+nums[0]
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 20)
+}
+
+func TestMutableParameterMultipleCalls(t *testing.T) {
+	// Test multiple calls with mutable primitive param
+	// Primitives now support true reference semantics with &
+	input := `
+do increment(&n int) {
+	n = n + 1
+}
+
+temp counter int = 0
+increment(counter)
+increment(counter)
+increment(counter)
+counter
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 3)
+}
+
+func TestMutableAndImmutableMixed(t *testing.T) {
+	// Test function with both mutable and immutable params
+	input := `
+do addToFirst(&arr [int], value int) {
+	arr[0] = arr[0] + value
+}
+
+temp nums [int] = {100, 200, 300}
+addToFirst(nums, 50)
+nums[0]
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 150)
+}
+
+func TestMutableNestedStructField(t *testing.T) {
+	// Test modifying nested struct field through & param
+	input := `
+const Address struct {
+	city string
+	zip int
+}
+
+const Person struct {
+	name string
+	addr Address
+}
+
+do updateZip(&p Person, newZip int) {
+	p.addr.zip = newZip
+}
+
+temp bob Person = Person{name: "Bob", addr: Address{city: "NYC", zip: 10001}}
+updateZip(bob, 90210)
+bob.addr.zip
+`
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 90210)
+}
+
+func TestSwapWithMutableParams(t *testing.T) {
+	// Test swap operation with two mutable primitive params
+	// Primitives now support true reference semantics with &
+	input := `
+do swap(&a, &b int) {
+	temp t int = a
+	a = b
+	b = t
+}
+
+temp x int = 10
+temp y int = 20
+swap(x, y)
+x + y * 100
+`
+	// Result should be 20 + 10*100 = 1020 if swap worked
+	// (x becomes 20, y becomes 10)
+	evaluated := testEval(input)
+	testIntegerObject(t, evaluated, 1020)
+}
+
+func TestMutableFloatParameter(t *testing.T) {
+	// Test mutable float parameter
+	input := `
+do setFloat(&f float) {
+	f = 3.14
+}
+
+temp val float = 0.0
+setFloat(val)
+val
+`
+	evaluated := testEval(input)
+	testFloatObject(t, evaluated, 3.14)
+}
+
+func TestMutableBoolParameter(t *testing.T) {
+	// Test mutable bool parameter
+	input := `
+do toggle(&b bool) {
+	b = !b
+}
+
+temp flag bool = false
+toggle(flag)
+flag
+`
+	evaluated := testEval(input)
+	testBooleanObject(t, evaluated, true)
+}
+
+func TestMutableStringParameter(t *testing.T) {
+	// Test mutable string parameter
+	input := `
+do setString(&s string) {
+	s = "modified"
+}
+
+temp str string = "original"
+setString(str)
+str
+`
+	evaluated := testEval(input)
+	testStringObject(t, evaluated, "modified")
+}

--- a/pkg/interpreter/object.go
+++ b/pkg/interpreter/object.go
@@ -37,6 +37,7 @@ type (
 	EnumValue       = object.EnumValue
 	Environment     = object.Environment
 	ModuleObject    = object.ModuleObject
+	Reference       = object.Reference
 )
 
 // Re-export constants
@@ -60,6 +61,7 @@ const (
 	ENUM_OBJ         = object.ENUM_OBJ
 	ENUM_VALUE_OBJ   = object.ENUM_VALUE_OBJ
 	MODULE_OBJ       = object.MODULE_OBJ
+	REFERENCE_OBJ    = object.REFERENCE_OBJ
 
 	// Visibility constants
 	VisibilityPublic        = object.VisibilityPublic

--- a/pkg/object/object.go
+++ b/pkg/object/object.go
@@ -34,6 +34,7 @@ const (
 	ENUM_VALUE_OBJ   ObjectType = "ENUM_VALUE"
 	MODULE_OBJ       ObjectType = "MODULE"
 	FILE_HANDLE_OBJ  ObjectType = "FILE_HANDLE"
+	REFERENCE_OBJ    ObjectType = "REFERENCE"
 )
 
 type Object interface {
@@ -111,6 +112,40 @@ func (fh *FileHandle) Inspect() string {
 		return fmt.Sprintf("<FileHandle(closed) %s>", fh.Path)
 	}
 	return fmt.Sprintf("<FileHandle %s>", fh.Path)
+}
+
+// Reference represents a reference to a variable in another environment
+// Used for mutable (&) parameters to allow modifications to persist to the caller
+type Reference struct {
+	Env  *Environment // The environment where the original variable lives
+	Name string       // The variable name in that environment
+}
+
+func (r *Reference) Type() ObjectType { return REFERENCE_OBJ }
+func (r *Reference) Inspect() string  { return fmt.Sprintf("<ref %s>", r.Name) }
+
+// Deref returns the current value of the referenced variable
+func (r *Reference) Deref() (Object, bool) {
+	return r.Env.Get(r.Name)
+}
+
+// SetValue updates the referenced variable's value
+// Traverses the scope chain to find and update the variable
+func (r *Reference) SetValue(val Object) bool {
+	return r.Env.updateRef(r.Name, val)
+}
+
+// updateRef updates a variable's value, traversing outer scopes if needed
+// This is used by Reference.SetValue to update through references
+func (e *Environment) updateRef(name string, val Object) bool {
+	if _, ok := e.store[name]; ok {
+		e.store[name] = val
+		return true
+	}
+	if e.outer != nil {
+		return e.outer.updateRef(name, val)
+	}
+	return false
 }
 
 // Boolean wraps bool

--- a/pkg/parser/parser.go
+++ b/pkg/parser/parser.go
@@ -1200,10 +1200,24 @@ func (p *Parser) parseFunctionParameters() []*Parameter {
 
 	p.nextToken()
 
+	// Helper struct to track name and mutability together
+	type paramInfo struct {
+		name    *Label
+		mutable bool
+	}
+
 	for {
 		// Collect parameter names that will share a type
 		// e.g., in "x, y int", collect ["x", "y"]
-		namesForType := []*Label{}
+		// e.g., in "&x, &y int", collect ["x", "y"] with mutable=true
+		namesForType := []paramInfo{}
+
+		// Check for & prefix (mutable parameter)
+		isMutable := false
+		if p.currentTokenMatches(AMPERSAND) {
+			isMutable = true
+			p.nextToken() // consume &
+		}
 
 		// Read first parameter name
 		if !p.currentTokenMatches(IDENT) {
@@ -1236,9 +1250,16 @@ func (p *Parser) parseFunctionParameters() []*Parameter {
 				} else {
 					paramNames[currentIdent.Value] = currentIdent.Token
 				}
-				namesForType = append(namesForType, currentIdent)
+				namesForType = append(namesForType, paramInfo{name: currentIdent, mutable: isMutable})
 				p.nextToken() // consume IDENT
-				p.nextToken() // consume COMMA, move to next IDENT
+				p.nextToken() // consume COMMA, move to next token
+
+				// Check for & prefix on next parameter in the shared-type group
+				isMutable = false
+				if p.currentTokenMatches(AMPERSAND) {
+					isMutable = true
+					p.nextToken() // consume &
+				}
 				continue
 			} else if p.peekTokenMatches(IDENT) || p.peekTokenMatches(LBRACKET) {
 				// This IDENT is a parameter name, and the next token is the type
@@ -1251,7 +1272,7 @@ func (p *Parser) parseFunctionParameters() []*Parameter {
 				} else {
 					paramNames[currentIdent.Value] = currentIdent.Token
 				}
-				namesForType = append(namesForType, currentIdent)
+				namesForType = append(namesForType, paramInfo{name: currentIdent, mutable: isMutable})
 				p.nextToken() // move to the type
 				break
 			} else if p.peekTokenMatches(RPAREN) {
@@ -1280,8 +1301,8 @@ func (p *Parser) parseFunctionParameters() []*Parameter {
 		}
 
 		// Apply the type to all collected names
-		for _, name := range namesForType {
-			params = append(params, &Parameter{Name: name, TypeName: typeName})
+		for _, info := range namesForType {
+			params = append(params, &Parameter{Name: info.name, TypeName: typeName, Mutable: info.mutable})
 		}
 
 		// Check for comma (more parameters) or closing paren

--- a/pkg/parser/parser_test.go
+++ b/pkg/parser/parser_test.go
@@ -1479,3 +1479,165 @@ func TestMatrixInStructField(t *testing.T) {
 		t.Errorf("expected field type '[[int]]', got %q", stmt.Fields[0].TypeName)
 	}
 }
+
+// ============================================================================
+// Mutable Parameter Tests (& prefix)
+// ============================================================================
+
+func TestMutableParameter(t *testing.T) {
+	input := `do modify(&p Person) {
+		p.age = 30
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 1 {
+		t.Fatalf("expected 1 parameter, got %d", len(fn.Parameters))
+	}
+	if fn.Parameters[0].Name.Value != "p" {
+		t.Errorf("expected parameter name 'p', got %q", fn.Parameters[0].Name.Value)
+	}
+	if fn.Parameters[0].TypeName != "Person" {
+		t.Errorf("expected parameter type 'Person', got %q", fn.Parameters[0].TypeName)
+	}
+	if !fn.Parameters[0].Mutable {
+		t.Errorf("expected parameter to be mutable (& prefix)")
+	}
+}
+
+func TestNonMutableParameter(t *testing.T) {
+	input := `do getName(p Person) -> string {
+		return p.name
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 1 {
+		t.Fatalf("expected 1 parameter, got %d", len(fn.Parameters))
+	}
+	if fn.Parameters[0].Mutable {
+		t.Errorf("expected parameter to be immutable (no & prefix)")
+	}
+}
+
+func TestMixedMutableParameters(t *testing.T) {
+	input := `do process(&arr [int], count int) {
+		arr[0] = count
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 2 {
+		t.Fatalf("expected 2 parameters, got %d", len(fn.Parameters))
+	}
+
+	// First param should be mutable
+	if fn.Parameters[0].Name.Value != "arr" {
+		t.Errorf("expected first param name 'arr', got %q", fn.Parameters[0].Name.Value)
+	}
+	if !fn.Parameters[0].Mutable {
+		t.Errorf("expected first parameter to be mutable")
+	}
+
+	// Second param should be immutable
+	if fn.Parameters[1].Name.Value != "count" {
+		t.Errorf("expected second param name 'count', got %q", fn.Parameters[1].Name.Value)
+	}
+	if fn.Parameters[1].Mutable {
+		t.Errorf("expected second parameter to be immutable")
+	}
+}
+
+func TestMultipleMutableParamsSharedType(t *testing.T) {
+	// Test: &a, &b int - both should be mutable with type int
+	input := `do swap(&a, &b int) {
+		temp t int = a
+		a = b
+		b = t
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 2 {
+		t.Fatalf("expected 2 parameters, got %d", len(fn.Parameters))
+	}
+
+	if fn.Parameters[0].Name.Value != "a" {
+		t.Errorf("expected first param name 'a', got %q", fn.Parameters[0].Name.Value)
+	}
+	if fn.Parameters[0].TypeName != "int" {
+		t.Errorf("expected first param type 'int', got %q", fn.Parameters[0].TypeName)
+	}
+	if !fn.Parameters[0].Mutable {
+		t.Errorf("expected first parameter to be mutable")
+	}
+
+	if fn.Parameters[1].Name.Value != "b" {
+		t.Errorf("expected second param name 'b', got %q", fn.Parameters[1].Name.Value)
+	}
+	if fn.Parameters[1].TypeName != "int" {
+		t.Errorf("expected second param type 'int', got %q", fn.Parameters[1].TypeName)
+	}
+	if !fn.Parameters[1].Mutable {
+		t.Errorf("expected second parameter to be mutable")
+	}
+}
+
+func TestMixedMutabilitySharedType(t *testing.T) {
+	// Test: &a, b int - a is mutable, b is immutable, both type int
+	input := `do test(&a, b int) {
+		a = b
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 2 {
+		t.Fatalf("expected 2 parameters, got %d", len(fn.Parameters))
+	}
+
+	if !fn.Parameters[0].Mutable {
+		t.Errorf("expected first parameter 'a' to be mutable")
+	}
+	if fn.Parameters[1].Mutable {
+		t.Errorf("expected second parameter 'b' to be immutable")
+	}
+}
+
+func TestMutableArrayParameter(t *testing.T) {
+	input := `do doubleAll(&arr [int]) {
+		arr[0] = arr[0] * 2
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 1 {
+		t.Fatalf("expected 1 parameter, got %d", len(fn.Parameters))
+	}
+	if fn.Parameters[0].TypeName != "[int]" {
+		t.Errorf("expected parameter type '[int]', got %q", fn.Parameters[0].TypeName)
+	}
+	if !fn.Parameters[0].Mutable {
+		t.Errorf("expected parameter to be mutable")
+	}
+}
+
+func TestMutableMapParameter(t *testing.T) {
+	input := `do addEntry(&m map[string:int], key string, val int) {
+		m[key] = val
+	}`
+	program := parseProgram(t, input)
+	fn := program.Statements[0].(*FunctionDeclaration)
+
+	if len(fn.Parameters) != 3 {
+		t.Fatalf("expected 3 parameters, got %d", len(fn.Parameters))
+	}
+	if fn.Parameters[0].TypeName != "map[string:int]" {
+		t.Errorf("expected first param type 'map[string:int]', got %q", fn.Parameters[0].TypeName)
+	}
+	if !fn.Parameters[0].Mutable {
+		t.Errorf("expected first parameter to be mutable")
+	}
+	if fn.Parameters[1].Mutable || fn.Parameters[2].Mutable {
+		t.Errorf("expected second and third parameters to be immutable")
+	}
+}

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -29,6 +29,7 @@ const (
 type Scope struct {
 	parent       *Scope
 	variables    map[string]string // variable name -> type name
+	mutability   map[string]bool   // variable name -> is mutable
 	usingModules map[string]bool   // modules imported via 'using'
 }
 
@@ -37,13 +38,32 @@ func NewScope(parent *Scope) *Scope {
 	return &Scope{
 		parent:       parent,
 		variables:    make(map[string]string),
+		mutability:   make(map[string]bool),
 		usingModules: make(map[string]bool),
 	}
 }
 
-// Define adds a variable to the current scope
+// Define adds a variable to the current scope (defaults to immutable)
 func (s *Scope) Define(name, typeName string) {
 	s.variables[name] = typeName
+	s.mutability[name] = false
+}
+
+// DefineWithMutability adds a variable to the current scope with explicit mutability
+func (s *Scope) DefineWithMutability(name, typeName string, mutable bool) {
+	s.variables[name] = typeName
+	s.mutability[name] = mutable
+}
+
+// IsMutable checks if a variable is mutable in the current scope or any parent scope
+func (s *Scope) IsMutable(name string) (bool, bool) {
+	if mutable, ok := s.mutability[name]; ok {
+		return mutable, true
+	}
+	if s.parent != nil {
+		return s.parent.IsMutable(name)
+	}
+	return false, false
 }
 
 // Lookup finds a variable in the current scope or any parent scope
@@ -119,8 +139,9 @@ type FunctionSignature struct {
 
 // Parameter represents a function parameter with type
 type Parameter struct {
-	Name string
-	Type string
+	Name    string
+	Type    string
+	Mutable bool // true if declared with & prefix
 }
 
 // TypeChecker validates types in an EZ program
@@ -485,8 +506,9 @@ func (tc *TypeChecker) checkFunctionDeclaration(node *ast.FunctionDeclaration) {
 			)
 		}
 		sig.Parameters = append(sig.Parameters, &Parameter{
-			Name: param.Name.Value,
-			Type: param.TypeName,
+			Name:    param.Name.Value,
+			Type:    param.TypeName,
+			Mutable: param.Mutable,
 		})
 	}
 
@@ -511,9 +533,9 @@ func (tc *TypeChecker) checkFunctionBody(node *ast.FunctionDeclaration) {
 	tc.enterScope()
 	defer tc.exitScope()
 
-	// Add function parameters to scope
+	// Add function parameters to scope with their mutability
 	for _, param := range node.Parameters {
-		tc.defineVariable(param.Name.Value, param.TypeName)
+		tc.defineVariableWithMutability(param.Name.Value, param.TypeName, param.Mutable)
 	}
 
 	// Check if function body contains at least one return statement (for functions with return types)
@@ -841,9 +863,9 @@ func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 		}
 	}
 
-	// Register variable in current scope
+	// Register variable in current scope with mutability (temp = mutable, const = immutable)
 	if declaredType != "" {
-		tc.defineVariable(varName, declaredType)
+		tc.defineVariableWithMutability(varName, declaredType, decl.Mutable)
 	}
 }
 
@@ -851,6 +873,20 @@ func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 func (tc *TypeChecker) checkAssignment(assign *ast.AssignmentStatement) {
 	// Also validate the value expression
 	tc.checkExpression(assign.Value)
+
+	// Check mutability - error if trying to modify an immutable variable (E5016)
+	if rootVar := tc.extractRootVariable(assign.Name); rootVar != "" {
+		isMutable, found := tc.isVariableMutable(rootVar)
+		if found && !isMutable {
+			line, column := tc.getExpressionPosition(assign.Name)
+			tc.addError(
+				errors.E5016,
+				fmt.Sprintf("cannot modify read-only parameter '%s'", rootVar),
+				line,
+				column,
+			)
+		}
+	}
 
 	// Get the target type
 	targetType, targetOk := tc.inferExpressionType(assign.Name)
@@ -882,6 +918,21 @@ func (tc *TypeChecker) checkAssignment(assign *ast.AssignmentStatement) {
 	// For index expressions, also validate the index
 	if indexExpr, ok := assign.Name.(*ast.IndexExpression); ok {
 		tc.checkIndexExpression(indexExpr)
+	}
+}
+
+// extractRootVariable returns the root variable name from an expression
+// For "x" returns "x", for "x.field" returns "x", for "arr[0]" returns "arr"
+func (tc *TypeChecker) extractRootVariable(expr ast.Expression) string {
+	switch e := expr.(type) {
+	case *ast.Label:
+		return e.Value
+	case *ast.MemberExpression:
+		return tc.extractRootVariable(e.Object)
+	case *ast.IndexExpression:
+		return tc.extractRootVariable(e.Left)
+	default:
+		return ""
 	}
 }
 
@@ -1291,7 +1342,7 @@ func (tc *TypeChecker) checkFunctionCall(call *ast.CallExpression) {
 		return
 	}
 
-	// Check argument types
+	// Check argument types and mutability
 	for i, arg := range call.Arguments {
 		actualType, ok := tc.inferExpressionType(arg)
 		if !ok {
@@ -1308,6 +1359,26 @@ func (tc *TypeChecker) checkFunctionCall(call *ast.CallExpression) {
 				line,
 				column,
 			)
+		}
+
+		// Check for const -> & param error (E3023)
+		// If the parameter is mutable (&), the argument must be a mutable variable
+		if sig.Parameters[i].Mutable {
+			// Check if argument is a simple variable (Label)
+			if label, isLabel := arg.(*ast.Label); isLabel {
+				// Check if this variable is mutable in scope
+				isMutable, found := tc.isVariableMutable(label.Value)
+				if found && !isMutable {
+					line, column := tc.getExpressionPosition(arg)
+					tc.addError(
+						errors.E3023,
+						fmt.Sprintf("cannot pass immutable variable '%s' to mutable parameter '&%s' in call to '%s'",
+							label.Value, sig.Parameters[i].Name, funcName),
+						line,
+						column,
+					)
+				}
+			}
 		}
 	}
 }
@@ -1676,11 +1747,26 @@ func (tc *TypeChecker) exitScope() {
 	}
 }
 
-// defineVariable adds a variable to the current scope
+// defineVariable adds a variable to the current scope (defaults to immutable)
 func (tc *TypeChecker) defineVariable(name, typeName string) {
 	if tc.currentScope != nil {
 		tc.currentScope.Define(name, typeName)
 	}
+}
+
+// defineVariableWithMutability adds a variable to the current scope with explicit mutability
+func (tc *TypeChecker) defineVariableWithMutability(name, typeName string, mutable bool) {
+	if tc.currentScope != nil {
+		tc.currentScope.DefineWithMutability(name, typeName, mutable)
+	}
+}
+
+// isVariableMutable checks if a variable is mutable in the current scope
+func (tc *TypeChecker) isVariableMutable(name string) (bool, bool) {
+	if tc.currentScope != nil {
+		return tc.currentScope.IsMutable(name)
+	}
+	return false, false
 }
 
 // lookupVariable finds a variable type in scope chain or global variables

--- a/pkg/typechecker/typechecker_test.go
+++ b/pkg/typechecker/typechecker_test.go
@@ -1040,3 +1040,177 @@ do main() {
 	tc := typecheck(t, input)
 	assertNoErrors(t, tc)
 }
+
+// ============================================================================
+// Mutable Parameter Tests (& prefix)
+// ============================================================================
+
+func TestMutableParameterValid(t *testing.T) {
+	// Test: temp variable passed to & param - should be valid
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do birthday(&p Person) {
+	p.age = p.age + 1
+}
+
+do main() {
+	temp bob Person = Person{name: "Bob", age: 30}
+	birthday(bob)
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestImmutableParameterValid(t *testing.T) {
+	// Test: const variable passed to non-& param - should be valid
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do getName(p Person) -> string {
+	return p.name
+}
+
+do main() {
+	const alice Person = Person{name: "Alice", age: 25}
+	temp name string = getName(alice)
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestE3023_ConstToMutableParam(t *testing.T) {
+	// Test: const variable passed to & param - should error E3023
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do modify(&p Person) {
+	p.age = p.age + 1
+}
+
+do main() {
+	const alice Person = Person{name: "Alice", age: 25}
+	modify(alice)
+}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E3023)
+}
+
+func TestE5016_ModifyImmutableParam(t *testing.T) {
+	// Test: modifying a non-& param inside function - should error E5016
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do tryModify(p Person) {
+	p.age = 100
+}
+
+do main() {
+	temp bob Person = Person{name: "Bob", age: 30}
+	tryModify(bob)
+}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E5016)
+}
+
+func TestE5016_ModifyImmutableParamField(t *testing.T) {
+	// Test: modifying a field on non-& param - should error E5016
+	input := `
+const Person struct {
+	name string
+	age int
+}
+
+do tryModifyField(p Person) {
+	p.name = "Changed"
+}
+
+do main() {
+	temp bob Person = Person{name: "Bob", age: 30}
+	tryModifyField(bob)
+}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E5016)
+}
+
+func TestE5016_ModifyImmutableArrayParam(t *testing.T) {
+	// Test: modifying array element on non-& param - should error E5016
+	input := `
+do tryModifyArray(arr [int]) {
+	arr[0] = 100
+}
+
+do main() {
+	temp nums [int] = {1, 2, 3}
+	tryModifyArray(nums)
+}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E5016)
+}
+
+func TestMutableArrayParamValid(t *testing.T) {
+	// Test: modifying array element on & param - should be valid
+	input := `
+do modifyArray(&arr [int]) {
+	arr[0] = 100
+}
+
+do main() {
+	temp nums [int] = {1, 2, 3}
+	modifyArray(nums)
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestMixedMutableParamsValid(t *testing.T) {
+	// Test: function with both mutable and immutable params
+	input := `
+do copyFirst(&dest [int], src [int]) {
+	dest[0] = src[0]
+}
+
+do main() {
+	temp target [int] = {0, 0, 0}
+	temp source [int] = {1, 2, 3}
+	copyFirst(target, source)
+}
+`
+	tc := typecheck(t, input)
+	assertNoErrors(t, tc)
+}
+
+func TestE3023_ConstArrayToMutableParam(t *testing.T) {
+	// Test: const array passed to & param - should error E3023
+	input := `
+do modify(&arr [int]) {
+	arr[0] = 100
+}
+
+do main() {
+	const nums [int] = {1, 2, 3}
+	modify(nums)
+}
+`
+	tc := typecheck(t, input)
+	assertHasError(t, tc, errors.E3023)
+}

--- a/tests/comprehensive.ez
+++ b/tests/comprehensive.ez
@@ -8,6 +8,7 @@
  *   - Control flow (if/or/otherwise, for, while, for_each, range)
  *   - All operators (arithmetic, comparison, logical, assignment)
  *   - Functions (parameters, returns, multiple returns)
+ *   - Mutable parameters (&) for pass-by-reference
  *   - Standard library (@std, @math, @arrays, @strings, @maps, @time)
  *   - String interpolation
  *   - Comments (single-line and multi-line)
@@ -129,6 +130,10 @@ do main() {
     total_failed += result.failed
 
     result = test_functions()
+    total_passed += result.passed
+    total_failed += result.failed
+
+    result = test_mutable_params()
     total_passed += result.passed
     total_failed += result.failed
 
@@ -742,6 +747,273 @@ do double_array(nums [int]) -> [int] {
 }
 
 // NOTE: min_max with multiple returns removed - not yet supported
+
+// ==================================================
+// MUTABLE PARAMETERS (&)
+// ==================================================
+
+do test_mutable_params() -> TestResult {
+    print_section("Mutable Parameters (&)")
+    temp passed int = 0
+    temp failed int = 0
+
+    // Basic mutable int parameter
+    temp counter int = 0
+    increment(counter)
+    if counter == 1 {
+        println("  increment(&n int): counter = ${counter} (expected 1)")
+        passed += 1
+    } otherwise {
+        println("  increment(&n int): FAILED - got ${counter}, expected 1")
+        failed += 1
+    }
+
+    // Multiple increments
+    increment(counter)
+    increment(counter)
+    if counter == 3 {
+        println("  multiple increments: counter = ${counter} (expected 3)")
+        passed += 1
+    } otherwise {
+        println("  multiple increments: FAILED - got ${counter}, expected 3")
+        failed += 1
+    }
+
+    // Swap two values
+    temp x int = 10
+    temp y int = 20
+    swap_values(x, y)
+    if x == 20 && y == 10 {
+        println("  swap(&a, &b int): x=${x}, y=${y} (expected x=20, y=10)")
+        passed += 1
+    } otherwise {
+        println("  swap(&a, &b int): FAILED - got x=${x}, y=${y}")
+        failed += 1
+    }
+
+    // Mutable float parameter
+    temp f float = 0.0
+    set_to_pi(f)
+    if f > 3.14 && f < 3.15 {
+        println("  set_to_pi(&f float): f = ${f} (expected ~3.14159)")
+        passed += 1
+    } otherwise {
+        println("  set_to_pi(&f float): FAILED - got ${f}")
+        failed += 1
+    }
+
+    // Mutable bool parameter
+    temp flag bool = false
+    toggle_bool(flag)
+    if flag == true {
+        println("  toggle_bool(&b bool): flag = ${flag} (expected true)")
+        passed += 1
+    } otherwise {
+        println("  toggle_bool(&b bool): FAILED - got ${flag}")
+        failed += 1
+    }
+    toggle_bool(flag)
+    if flag == false {
+        println("  toggle again: flag = ${flag} (expected false)")
+        passed += 1
+    } otherwise {
+        println("  toggle again: FAILED - got ${flag}")
+        failed += 1
+    }
+
+    // Mutable string parameter
+    temp message string = "Hello"
+    append_world(message)
+    if message == "Hello, World!" {
+        println("  append_world(&s string): '${message}'")
+        passed += 1
+    } otherwise {
+        println("  append_world(&s string): FAILED - got '${message}'")
+        failed += 1
+    }
+
+    // Mutable byte parameter
+    temp b byte = 0x0F
+    shift_byte_left(b)
+    if b == 0x1E {
+        println("  shift_byte_left(&b byte): b = ${b} (expected 30/0x1E)")
+        passed += 1
+    } otherwise {
+        println("  shift_byte_left(&b byte): FAILED - got ${b}")
+        failed += 1
+    }
+
+    // Mutable byte array parameter
+    temp bytes [byte] = {0x01, 0x02, 0x03}
+    double_bytes(bytes)
+    if bytes[0] == 0x02 && bytes[1] == 0x04 && bytes[2] == 0x06 {
+        println("  double_bytes(&arr [byte]): ${bytes}")
+        passed += 1
+    } otherwise {
+        println("  double_bytes(&arr [byte]): FAILED - got ${bytes}")
+        failed += 1
+    }
+
+    // Mutable struct parameter
+    temp p Point = Point{x: 5, y: 10}
+    double_point(p)
+    if p.x == 10 && p.y == 20 {
+        println("  double_point(&p Point): x=${p.x}, y=${p.y} (expected 10, 20)")
+        passed += 1
+    } otherwise {
+        println("  double_point(&p Point): FAILED - got x=${p.x}, y=${p.y}")
+        failed += 1
+    }
+
+    // Mutable array parameter
+    temp nums [int] = {1, 2, 3, 4, 5}
+    double_array_in_place(nums)
+    if nums[0] == 2 && nums[2] == 6 && nums[4] == 10 {
+        println("  double_array_in_place(&arr [int]): ${nums}")
+        passed += 1
+    } otherwise {
+        println("  double_array_in_place(&arr [int]): FAILED - got ${nums}")
+        failed += 1
+    }
+
+    // Mixed mutable and immutable parameters
+    temp result int = 0
+    add_to_result(result, 15)
+    add_to_result(result, 25)
+    if result == 40 {
+        println("  add_to_result(&result, value): result = ${result} (expected 40)")
+        passed += 1
+    } otherwise {
+        println("  add_to_result(&result, value): FAILED - got ${result}")
+        failed += 1
+    }
+
+    // Multiple output pattern (quotient and remainder)
+    temp q int = 0
+    temp r int = 0
+    divide_with_remainder(q, r, 17, 5)
+    if q == 3 && r == 2 {
+        println("  divide_with_remainder: 17/5 = ${q} remainder ${r}")
+        passed += 1
+    } otherwise {
+        println("  divide_with_remainder: FAILED - got q=${q}, r=${r}")
+        failed += 1
+    }
+
+    // Accumulator pattern with string
+    temp greeting string = ""
+    append_name(greeting, "Alice")
+    append_name(greeting, "Bob")
+    append_name(greeting, "Carol")
+    temp expected_greeting string = "Hello, Alice! Hello, Bob! Hello, Carol! "
+    if greeting == expected_greeting {
+        println("  accumulator pattern: '${greeting}'")
+        passed += 1
+    } otherwise {
+        println("  accumulator pattern: FAILED")
+        println("    got:      '${greeting}'")
+        println("    expected: '${expected_greeting}'")
+        failed += 1
+    }
+
+    // State machine pattern (game state)
+    temp game GameState = GameState{score: 0, lives: 3, level: 1}
+    add_score(game, 50)
+    add_score(game, 60)  // This should trigger level up (score >= level * 100)
+    if game.score == 110 && game.level == 2 {
+        println("  state machine: score=${game.score}, level=${game.level}")
+        passed += 1
+    } otherwise {
+        println("  state machine: FAILED - score=${game.score}, level=${game.level}")
+        failed += 1
+    }
+
+    lose_life(game)
+    if game.lives == 2 {
+        println("  lose_life: lives=${game.lives} (expected 2)")
+        passed += 1
+    } otherwise {
+        println("  lose_life: FAILED - lives=${game.lives}")
+        failed += 1
+    }
+
+    println("  PASSED: ${passed}, FAILED: ${failed}")
+    return TestResult{passed: passed, failed: failed}
+}
+
+// Helper functions for mutable parameter tests
+do increment(&n int) {
+    n = n + 1
+}
+
+do swap_values(&a, &b int) {
+    temp t int = a
+    a = b
+    b = t
+}
+
+do set_to_pi(&f float) {
+    f = 3.14159
+}
+
+do toggle_bool(&b bool) {
+    b = !b
+}
+
+do append_world(&s string) {
+    s = s + ", World!"
+}
+
+do shift_byte_left(&b byte) {
+    b = b * 2
+}
+
+do double_bytes(&arr [byte]) {
+    for i in range(0, len(arr)) {
+        arr[i] = arr[i] * 2
+    }
+}
+
+do double_point(&p Point) {
+    p.x = p.x * 2
+    p.y = p.y * 2
+}
+
+do double_array_in_place(&arr [int]) {
+    for i in range(0, len(arr)) {
+        arr[i] = arr[i] * 2
+    }
+}
+
+do add_to_result(&result int, value int) {
+    result = result + value
+}
+
+do divide_with_remainder(&quotient, &remainder int, a, b int) {
+    quotient = a / b
+    remainder = a % b
+}
+
+do append_name(&message string, name string) {
+    message = message + "Hello, " + name + "! "
+}
+
+const GameState struct {
+    score int
+    lives int
+    level int
+}
+
+do add_score(&state GameState, points int) {
+    state.score = state.score + points
+    if state.score >= state.level * 100 {
+        state.level = state.level + 1
+    }
+}
+
+do lose_life(&state GameState) {
+    state.lives = state.lives - 1
+}
 
 // ==================================================
 // STDLIB: @math

--- a/tests/errors/comprehensive/E3023_const_to_mutable_param.ez
+++ b/tests/errors/comprehensive/E3023_const_to_mutable_param.ez
@@ -1,0 +1,18 @@
+/*
+ * Error Test: E3023 - const-to-mutable-param
+ * Expected: "cannot pass immutable variable" to mutable parameter
+ */
+
+const Person struct {
+    name string
+    age int
+}
+
+do modify(&p Person) {
+    p.age = p.age + 1
+}
+
+do main() {
+    const alice Person = Person{name: "Alice", age: 25}
+    modify(alice)  // ERROR: cannot pass const to & param
+}

--- a/tests/errors/comprehensive/E5016_immutable_parameter.ez
+++ b/tests/errors/comprehensive/E5016_immutable_parameter.ez
@@ -1,0 +1,18 @@
+/*
+ * Error Test: E5016 - immutable-parameter
+ * Expected: "cannot modify read-only parameter"
+ */
+
+const Person struct {
+    name string
+    age int
+}
+
+do try_modify(p Person) {
+    p.age = p.age + 1  // ERROR: cannot modify non-& param
+}
+
+do main() {
+    temp bob Person = Person{name: "Bob", age: 30}
+    try_modify(bob)
+}

